src/libraries/System.Private.Uri/src/System/Uri.cs

```diff
@@ -126,8 +140,19 @@ private class UriInfo
             public string? String;
             public Offset Offset;
             public string? DnsSafeHost;    // stores dns safe host when idn is on and we have unicode or idn host
-            public MoreInfo? MoreInfo;     // Multi-threading: This field must be always accessed through a _local_
-                                           // stack copy of m_Info.
+
+            private MoreInfo? _moreInfo;
+            public MoreInfo MoreInfo
+            {
+                get
+                {
+                    if (_moreInfo is null)
+                    {
+                        Interlocked.CompareExchange(ref _moreInfo, new MoreInfo(), null);
+                    }
+                    return _moreInfo;
+                }
+            }
```

Comment:
While there used to be a comment here to be careful about accesing this field, the code wasn't careful.
There were race conditions where we could end up reseting the `MoreInfo` field, leading to incorrect results being returned by `Equals`